File: services/prediction-service/app/schemas/analysis.py
from datetime import UTC, datetime
from enum import StrEnum
from typing import Literal, Self

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    HttpUrl,
    field_validator,
    model_validator,
)


class ContractModel(BaseModel):
    model_config = ConfigDict(
        populate_by_name=True,
        extra="forbid",
        str_strip_whitespace=True,
    )


class AnalysisDomain(StrEnum):
    GENERAL_RESEARCH = "general_research"
    CUSTOM_DATASET = "custom_dataset"
    SPORTS = "sports"
    FINANCIAL_MARKET = "financial_market"


class AnalysisStatus(StrEnum):
    PENDING = "pending"
    QUEUED = "queued"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class ConfidenceLevel(StrEnum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"


class EvidenceImpact(StrEnum):
    SUPPORTS = "SUPPORTS"
    OPPOSES = "OPPOSES"
    NEUTRAL = "NEUTRAL"


class StrengthLevel(StrEnum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"


class FreshnessStatus(StrEnum):
    CURRENT = "CURRENT"
    AGING = "AGING"
    STALE = "STALE"
    UNKNOWN = "UNKNOWN"


class AnalysisOptions(ContractModel):
    include_explanation: bool | None = Field(
        default=None,
        alias="includeExplanation",
    )
    include_sources: bool | None = Field(
        default=None,
        alias="includeSources",
    )
    include_confidence: bool | None = Field(
        default=None,
        alias="includeConfidence",
    )
    time_horizon: str | None = Field(
        default=None,
        alias="timeHorizon",
    )
    risk_tolerance: Literal["low", "medium", "high"] | None = Field(
        default=None,
        alias="riskPreference",
    )


class AnalysisServiceRequest(ContractModel):
    analysis_id: str = Field(
        alias="analysisId",
        min_length=1,
    )
    question: str = Field(
        min_length=3,
        max_length=5000,
    )
    domain: AnalysisDomain
    options: AnalysisOptions | None = None
    dataset_object_key: str | None = Field(
        default=None,
        alias="datasetObjectKey",
    )
    correlation_id: str = Field(
        alias="correlationId",
        min_length=1,
    )


class AnalysisConfidence(ContractModel):
    score: float | None = Field(ge=0, le=1)
    level: ConfidenceLevel | None
    explanation: str | None

    @model_validator(mode="after")
    def validate_score_and_level(self) -> Self:
        if (self.score is None) != (self.level is None):
            raise ValueError(
                "confidence score and level must either both be set or both be null"
            )

        return self


class EvidenceItem(ContractModel):
    id: str = Field(min_length=1)
    title: str = Field(min_length=1)
    description: str = Field(min_length=1)
    impact: EvidenceImpact | None
    strength: StrengthLevel | None


class RiskFactor(ContractModel):
    id: str = Field(min_length=1)
    title: str = Field(min_length=1)
    description: str = Field(min_length=1)
    severity: StrengthLevel | None


class AnalysisSource(ContractModel):
    id: str = Field(min_length=1)
    title: str = Field(min_length=1)
    url: HttpUrl | None
    publisher: str | None
    retrieved_at: datetime | None = Field(alias="retrievedAt")

    @field_validator("retrieved_at")
    @classmethod
    def validate_retrieved_at(
        cls,
        value: datetime | None,
    ) -> datetime | None:
        if value is None:
            return None

        if value.tzinfo is None or value.utcoffset() is None:
            raise ValueError("retrievedAt must include a timezone")

        return value.astimezone(UTC)


class ModelInformation(ContractModel):
    name: str = Field(min_length=1)
    version: str = Field(min_length=1)
    method: str | None


class DataFreshness(ContractModel):
    generated_at: datetime = Field(alias="generatedAt")
    data_as_of: datetime | None = Field(alias="dataAsOf")
    status: FreshnessStatus

    @field_validator("generated_at")
    @classmethod
    def validate_generated_at(cls, value: datetime) -> datetime:
        if value.tzinfo is None or value.utcoffset() is None:
            raise ValueError("generatedAt must include a timezone")

        return value.astimezone(UTC)

    @field_validator("data_as_of")
    @classmethod
    def validate_data_as_of(
        cls,
        value: datetime | None,
    ) -> datetime | None:
        if value is None:
            return None

        if value.tzinfo is None or value.utcoffset() is None:
            raise ValueError("dataAsOf must include a timezone")

        return value.astimezone(UTC)


class AnalysisResult(ContractModel):
    direct_answer: str = Field(
        alias="directAnswer",
        min_length=1,
    )
    probability: float | None = Field(ge=0, le=1)
    confidence: AnalysisConfidence
    evidence: list[EvidenceItem] = Field(default_factory=list)
    risk_factors: list[RiskFactor] = Field(
        default_factory=list,
        alias="riskFactors",
    )
    suggested_action: str | None = Field(alias="suggestedAction")
    sources: list[AnalysisSource] = Field(default_factory=list)
    model: ModelInformation
    data_freshness: DataFreshness = Field(alias="dataFreshness")


class AnalysisError(ContractModel):
    code: str = Field(min_length=1)
    message: str = Field(min_length=1)


class AnalysisServiceResponse(ContractModel):
    analysis_id: str = Field(
        alias="analysisId",
        min_length=1,
    )
    status: AnalysisStatus
    result: AnalysisResult | None
    processing_time_ms: float | None = Field(
        alias="processingTimeMs",
        ge=0,
    )
    error: AnalysisError | None

    @model_validator(mode="after")
    def validate_status_payload(self) -> Self:
        if self.status == AnalysisStatus.COMPLETED and self.result is None:
            raise ValueError("A completed analysis must contain a result")

        if self.status == AnalysisStatus.FAILED and self.error is None:
            raise ValueError("A failed analysis must contain an error")

        if self.status == AnalysisStatus.FAILED and self.result is not None:
            raise ValueError("A failed analysis cannot contain a result")

        if self.status != AnalysisStatus.FAILED and self.error is not None:
            raise ValueError("Only a failed analysis can contain an error")

        return self

File: packages/contracts/src/analysis.ts
/**
 * Canonical ForecastMe analysis domains.
 *
 * Provider platforms such as Polymarket, Kalshi, and sportsbooks are
 * contexts or data providers, not top-level analysis domains.
 */
export const AnalysisDomain = {
  GENERAL_RESEARCH: 'GENERAL_RESEARCH',
  CUSTOM_DATASET: 'CUSTOM_DATASET',
  SPORTS: 'SPORTS',
  FINANCIAL_MARKET: 'FINANCIAL_MARKET',
} as const;

export type AnalysisDomain = (typeof AnalysisDomain)[keyof typeof AnalysisDomain];

/**
 * Analytical task detected from the user's request.
 */
export const ClassificationTask = {
  GENERAL_RESEARCH: 'GENERAL_RESEARCH',
  DATASET_ANALYSIS: 'DATASET_ANALYSIS',
  OUTCOME_FORECAST: 'OUTCOME_FORECAST',
  DIRECTIONAL_FORECAST: 'DIRECTIONAL_FORECAST',
  COMPARISON: 'COMPARISON',
  RISK_ASSESSMENT: 'RISK_ASSESSMENT',
  UNSUPPORTED: 'UNSUPPORTED',
} as const;

export type ClassificationTask = (typeof ClassificationTask)[keyof typeof ClassificationTask];

/**
 * Classifier that produced the classification metadata.
 */
export const ClassifierSource = {
  LLM: 'LLM',
  RULE_BASED_FALLBACK: 'RULE_BASED_FALLBACK',
} as const;

export type ClassifierSource = (typeof ClassifierSource)[keyof typeof ClassifierSource];

/**
 * Canonical lifecycle state of an analysis.
 */
export const AnalysisStatus = {
  PENDING: 'PENDING',
  CLASSIFYING: 'CLASSIFYING',
  COLLECTING_DATA: 'COLLECTING_DATA',
  ANALYZING: 'ANALYZING',
  COMPLETED: 'COMPLETED',
  FAILED: 'FAILED',
} as const;

export type AnalysisStatus = (typeof AnalysisStatus)[keyof typeof AnalysisStatus];

/**
 * User-configurable analysis parameters.
 */
export interface AnalysisOptions {
  includeExplanation?: boolean;
  includeSources?: boolean;
  includeConfidence?: boolean;
  timeHorizon?: string;
  riskPreference?: 'low' | 'medium' | 'high';
}

/**
 * Metadata produced when the Python service classifies an analysis request.
 *
 * Unsupported requests still use one of the four canonical domains and are
 * identified through isSupported instead of introducing another domain.
 */
export interface ClassificationMetadata {
  domain: AnalysisDomain;
  task: ClassificationTask;
  confidence: number;
  reasoning: string;
  isSupported: boolean;
  entities: string[];
  dates: string[];
  timeHorizon: string | null;
  requiresLiveData: boolean;
  classifier: ClassifierSource;
  predictionIntent: boolean;
  comparisonIntent: boolean;
  riskIntent: boolean;
}

/**
 * Public request sent by the frontend to the NestJS Core API.
 */
export interface CreateAnalysisRequest {
  question: string;
  domain: AnalysisDomain;
  options?: AnalysisOptions;
  datasetId?: string;
}

/**
 * Internal request sent by the NestJS Core API to the Python Analysis Service.
 *
 * User-sensitive data should not be added unless analytically necessary.
 */
export interface AnalysisServiceRequest {
  analysisId: string;
  question: string;
  domain: AnalysisDomain;
  options?: AnalysisOptions;
  datasetObjectKey?: string;
  correlationId: string;
}

/**
 * Supported confidence levels.
 */
export const confidenceLevels = ['LOW', 'MEDIUM', 'HIGH'] as const;

export type ConfidenceLevel = (typeof confidenceLevels)[number];

/**
 * How an evidence item affects the predicted outcome.
 */
export const evidenceImpacts = ['SUPPORTS', 'OPPOSES', 'NEUTRAL'] as const;

export type EvidenceImpact = (typeof evidenceImpacts)[number];

/**
 * Supported evidence-strength and risk-severity levels.
 */
export const strengthLevels = ['LOW', 'MEDIUM', 'HIGH'] as const;

export type StrengthLevel = (typeof strengthLevels)[number];

/**
 * Supported data-freshness states.
 */
export const freshnessStatuses = ['CURRENT', 'AGING', 'STALE', 'UNKNOWN'] as const;

export type FreshnessStatus = (typeof freshnessStatuses)[number];

/**
 * Confidence describes how reliable ForecastMe considers an estimate.
 *
 * It is separate from probability, which describes the estimated likelihood
 * of an outcome.
 */
export interface AnalysisConfidence {
  score: number | null;
  level: ConfidenceLevel | null;
  explanation: string | null;
}

/**
 * Verified information used to support or oppose the result.
 */
export interface EvidenceItem {
  id: string;
  title: string;
  description: string;
  impact: EvidenceImpact | null;
  strength: StrengthLevel | null;
}

/**
 * A condition that could negatively affect the result.
 */
export interface RiskFactor {
  id: string;
  title: string;
  description: string;
  severity: StrengthLevel | null;
}

/**
 * A verified source used during the analysis.
 */
export interface AnalysisSource {
  id: string;
  title: string;
  url: string | null;
  publisher: string | null;
  retrievedAt: string | null;
}

/**
 * Exact model or analytical method used to produce the result.
 */
export interface ModelInformation {
  name: string;
  version: string;
  method: string | null;
}

/**
 * Timestamp and freshness information for the result.
 *
 * Timestamps must be UTC ISO 8601 strings.
 */
export interface DataFreshness {
  generatedAt: string;
  dataAsOf: string | null;
  status: FreshnessStatus;
}

/**
 * Canonical ForecastMe result shared by Python, NestJS, persistence,
 * and the frontend.
 *
 * Probability uses a zero-to-one scale and must be null when it cannot
 * be calculated. The frontend is responsible for percentage formatting.
 */
export interface AnalysisResult {
  directAnswer: string;
  probability: number | null;
  confidence: AnalysisConfidence;
  evidence: EvidenceItem[];
  riskFactors: RiskFactor[];
  suggestedAction: string | null;
  sources: AnalysisSource[];
  model: ModelInformation;
  dataFreshness: DataFreshness;
}

/**
 * Structured response returned by the Python Analysis Service.
 */
export interface AnalysisServiceResponse {
  analysisId: string;
  status: AnalysisStatus;
  result: AnalysisResult | null;
  processingTimeMs: number | null;
  error: {
    code: string;
    message: string;
  } | null;
}

/**
 * Public analysis object returned by the NestJS API.
 */
export interface AnalysisResponse {
  id: string;
  question: string;
  domain: AnalysisDomain;
  status: AnalysisStatus;
  result: AnalysisResult | null;
  createdAt: string;
  updatedAt: string;
  completedAt: string | null;
}

/**
 * Lightweight analysis entry used in history lists.
 */
export interface AnalysisHistoryItem {
  id: string;
  question: string;
  domain: AnalysisDomain;
  status: AnalysisStatus;
  probability: number | null;
  confidenceLevel: ConfidenceLevel | null;
  createdAt: string;
  completedAt: string | null;
}
